### ISSUE TITLE
Worker Pools

### ISSUE THREAD

**Issue by droyad** [OctopusDeploy/Specs#30]
Place Holder. More detail to be added

**Comment by droyad** [comment]
The simplest option is to have Worker Pools:
- The step would choose which pool it wants to run in
- A worker could be in multiple pools

This would involve no detection logic, we just trust the user has set it up right.

I find the TeamCity approach of on-image property configuration and requiring the image to be spun up cumbersome.

**Comment by tothegills** [comment]
**TLDR:**
Worker1 is a Tentacle worker with the role `Azure 3.0` in the environments `Test`, `Staging`
Worker2 is a Tentacle worker with the role `Azure 3.0` in the environments `Test`, `Staging`
Worker3 is a Tentacle worker with the role `Azure 3.0` in the environment `Production`
Worker4 is an SSH worker with the role `AWS` in the environment `Production`

Step 1 runs on an `Azure 3.0` worker (Worker1 or Worker2 in `Test` and `Staging`, Worker3 in `Production`) 

---
Kinda sounds like we have this concept already, except we just want "run on one" instead of "run on all".
Workers with different capabilities are deployment targets and their capabilities are defined by roles (`Azure 3.0` has the Azure SDK, `AWS` has the aws cli). Pools of workers are bit like environments.

Maybe deployment targets can be anointed "worker" which will change how they are involved in a deployment: excluded from "run on target" steps, and only one chosen for "run on worker" steps. Is there a need for pools, or will roles/environments work? ie Production has a dedicated worker that is locked down to the max, all the other environments share two workers. The worker pool for a step is defined by matching roles and environment.

We could create new steps to deploy to workers.  We recycled the "run script" step because people already had a Tentacle installed on their Octopus Server and it was easy for them to migrate to "run on server" with the flick of a toggle. On the downside, the target definition is complicated by a big OR that completely changes how the step works. "This step runs on all deployment targets or runs once on the octopus server (except when run on the server on behalf of deployment targets)"

With dedicated worker steps I think it would be clearer: "This step runs on an `Azure 3.0` worker" or "This step runs on an `Azure 3.0` worker in the context of each deployment target with the `web-app` role." The implementation would also be cleaner, we could do away with the `if (runningOnServer) ... else ...`.

**Comment by matt-richardson** [comment]
Worker pools are necessary to avoid having single points of failure. When "the only path to prod" is via octopus, you dont want to have a single worker going AWOL to break your deployments.

I'd be hesitant to overload environments/roles - it feels like it would make things much harder to grok.

**Comment by tothegills** [comment]
I get that we don't want a single point of failure, but I'm wondering if workers need a whole new vernacular. I don't think environment and roles are being overloaded, they have the same meaning. I would argue it's easier to grok because we would use the same patterns that people need to know to perform deployments anyway. The only distinction would be that steps run on all deployment targets, but only run on one worker. You can have as many workers as you want with whatever roles you want in whatever environments you want. The pool is the set of possible workers that a step can run on. We can derive that pool the same way we determine deployment targets, using role/environment.

**Comment by zentron** [comment]
As a note to this some of the ideas under the `resources` banner involve changing how roles work and instead allowing targets to be tagged. one tag could be called `Role` and it would function like current roles, but then the user may want to add `Framework` that, during say a health check is auto populated with `.Net3.5` or whatever. Deployments the would use a set of rules (which define a resource) to resolve the appropriate machine(s).  I think that may cover\assist this particular feature.
